source/fab/steps/grab/svn.py

```diff
@@ -3,32 +3,160 @@
+            If not specified, the code is copied into the root of the source folder.
+        :param revision:
+            E.g 'vn6.3'.
+        :param name:
+            Human friendly name for logger output, with sensible default.
+
+        """
+        # pull the revision out of the url, if it's in there
+        src, revision = _get_revision(src, revision)
+        name = name or f'{self.__class__.__name__} {dst} {revision}'.strip()
+        super().__init__(src, dst, name=name, revision=revision)
+
+    def run(self, artefact_store: Dict, config):
+        if not self.tool_available():
+            raise RuntimeError(f"command line tool not available: '{self.command}'")
+        super().run(artefact_store, config)
+
+    @classmethod
+    def tool_available(cls) -> bool:
+        """Is the command line tool available?"""
+        try:
+            run_command([cls.command, 'help'])
+        except FileNotFoundError:
+            return False
+        return True
+
+    def _cli_revision_parts(self):
+        # return the command line argument to specif the revision, if there is one
+        return ['--revision', str(self.revision)] if self.revision is not None else []
 
-    You can include a branch in the URL, for example:
+    def _is_working_copy(self, dst: Union[str, Path]) -> bool:
+        # is the given path is a working copy?
+        try:
+            run_command([self.command, 'info'], cwd=dst)
+        except RuntimeError:
+            return False
+        return True
 
-        GrabSvn(
-            src='https://code.metoffice.gov.uk/svn/lfric/GPL-utilities/trunk',
-            revision=36615, dst='gpl_utils')
+
+class SvnExport(GrabSvnBase):
+    """
+    Export an FCM repo folder to the project workspace.
 
     """
     def run(self, artefact_store: Dict, config):
         super().run(artefact_store, config)
 
-        if not svn:
-            raise ImportError('svn not installed, unable to continue')
+        run_command([
+            self.command, 'export', '--force',
+            *self._cli_revision_parts(),
+            self.src,
+            str(self._dst)
+        ])
+
+
+class SvnCheckout(GrabSvnBase):
+    """
+    Checkout or update an FCM repo.
+
+    .. note::
+        If the destination is a working copy, it will be updated to the given revision, **ignoring the source url**.
+        As such, the revision should be provided via the argument, not as part of the url.
+
+    """
+    def run(self, artefact_store: Dict, config):
+        super().run(artefact_store, config)
+
+        # new folder?
+        if not self._dst.exists():  # type: ignore
+            run_command([
+                self.command, 'checkout',
+                *self._cli_revision_parts(),
+                self.src, str(self._dst)
+            ])
+
+        else:
+            # working copy?
+            if self._is_working_copy(self._dst):  # type: ignore
```

Comment:
If `_is_working_copy(...)` is called externally (including from a child) then it isn't "private".

Reply:
In Python, a _double_ preceding underscore indicates private members and invokes name mangling. A _single_ preceding underscore indicates protected members, does not invoke name mangling and is accessible from subclasses.

This code:
```
class Base(object):
    def __private_func(self):
        print("subclasses can't call me")
    def _protected_func(self):
        print("subclasses can call me")

class Derrived(Base):
    pass

if __name__ == '__main__':
    Derrived()._protected_func()
    Derrived().__private_func()
```

produces this output:
```
subclasses can call me
Traceback (most recent call last):
  File "/home/h02/bblay/git/fab/temp/private_vs_protected.py", line 17, in <module>
    Derrived().__private_func()
AttributeError: 'Derrived' object has no attribute '__private_func'. Did you mean: '_Base__private_func'?
```

Reply:
I did not know about the double underscore thing although upon researching it I see it was introduced in 2001 with PEP8. Unfortunately this doesn't change the problem here as it is rarely a good idea to use protected. In fact the developers of C++ consider its existence to be their biggest blunder.

If a child needs to access a member variable then it should be through an accessor (property is fine) and never directly regardless of whether it is possible.

Directly accessing parent variables means you have to know what is in the parent and what consistency requirements it has. Using an accessor means you only need to know the API and consistency is maintained for free.

Reply:
That's an interesting point. I'll make the change, no problem, although this isn't a member variable, it's a method, so the solution would presumably be to make it public? That's a bit weird because we don't really need other classes to call this. Shall I go ahead with that anyway?

Just because it's an interesting topic I love to discuss:
I disagree that Python, C++, Java (etc?) were wrong to add protected variables, and I disagree that access by a subclass breaks encapsulation. You don't have to humor me with a reply, or perhaps we could have a nerdy coffee break sometime.

Reply:
In my zeal for eliminating the use of non-public members I overlooked that it was a method rather than a property. Still, is there any reason why it can't be public? Is there any reason why someone holding one of these objects shouldn't know if it is associated with a working copy?

Reply:
I made it private to keep the interface small but there's no harm in making it public, will do.